crates/matrix-sdk/src/room/common.rs

```diff
@@ -99,22 +104,68 @@ impl Common {
 
     /// Leave this room.
     ///
+    /// Returns a [`Result`] containing an instance of [`Left`] if successful.
+    ///
     /// Only invited and joined rooms can be left
-    pub(crate) async fn leave(&self) -> Result<()> {
+    #[doc = concat!("\n\n", include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/src/docs/sync_running.md")))]
+    pub(crate) async fn leave(&self) -> Result<Left> {
         let request = leave_room::v3::Request::new(self.inner.room_id());
-        let _response = self.client.send(request, None).await?;
 
-        Ok(())
+        let (tx, mut rx) = mpsc::channel::<Result<Left>>(1);
+
+        let user_id = if let Some(user_id) = self.client.user_id() {
+            user_id.to_owned()
+        } else {
+            return Err(Error::AuthenticationRequired);
+        };
+
+        let handle = self.add_event_handler({
+            move |event: SyncStateEvent<RoomMemberEventContent>, room: Room| {
+                let mut tx = tx.clone();
+                let user_id = user_id.clone();
+
+                async move {
+                    if (event.membership() == &MembershipState::Leave)
+                        && (event.state_key() == &user_id)
```

Comment:
I don't think these parentheses are needed (for readability or otherwise). Also I think it's more common to deref one side of the comparison rather than adding an extra reference to the other when types don't match up.